### ISSUE TITLE
Simulation friendly pressurisation method (suggestion)

### ISSUE THREAD

**Issue by cloudrambler** [tsunamayo/Starship-EVO#2907]
Starship-EVO is a fun oriented game and keeping it that way is great. However, being in space should also feel risky. Moreover, being in your nice cosy ship should feel safer than being outside in the vacuum. Such a hazard would add risk to balance out the reward of having a good ship available. So a pressurisation mechanic would add a lot of depth.

I know Space Engineers had this mechanic early on and then removed it due to simulation cost problems. So what I propose is a very much simpler method of implementing a functional pressurisation mechanic that is much lighter on the simulation.

Rather than asking the computer to prove if a given internal volume is enclosed at all points, we could do the task manually in advance. For a given room, which could be defined by the presence of a special vent brick (for example), we associate all of the surrounding blocks and bricks that we think best define that room curtilage. The simulation then only has to verify this once, that the chosen curtilage is in fact complete for that vent brick. Then if any one of those associated blocks or bricks is damaged, the room is assumed to lose pressure (regardless of any other surrounding structure).

With that in place the vent brick can control whether the room is pressurised or not. Open doors would also have to be considered, but these would be defined initially, thus linking one vent brick control area to another.

The case of user made doors would also work if the determination of damage also included temporary relocation.

Players that did not want to use this mechanic could instead place a single (magic) block that would grant them breathable air provided they stay within the collision sphere of that ship.

**Comment by cloudrambler** [comment]
Update: when determining if a curtilage is complete, the game could tolerate small gaps. This would allow imperfect seals, such as with angled ramp doors, to be considered as sealed for the purpose of pressurisation. For such tricky arrangements, the argument for those blocks being displaced could be extended to include the activation of any supporting mechanism. For example, as soon as a tail ramp rotor is powered, the ramp is assumed to be removed and the room is assumed to have lost pressure. Thus the simulation does not need to consider the actual gap or how it may change with the actual movement.

In short, a pressurisation mechanic that uses worst case assumptions, is still better than none. Especially if we can choose which blocks define the curtilage and thus have control over how robust each room would be treated by the simulation.

**Comment by tsunamayo** [comment]
Thanks for your suggestion! Air pressure is on the roadmap indeed. Your implementation seems a very complex to do, also there is a lot of other issue you have to take into account. My implementation is voxel based (only hull and blast-door will be airtight), and it will be multithreaded / run on compute shader for performance.
Cheers

**Comment by cloudrambler** [comment]
@tsunamayo Thanks for the explanation. Can you confirm if your intention is to limit this method (and general room delimitation for AI) to blocks, of will some bricks be factored into the ship function mapping? Cheers.
